ui/src/app/views/profile-page/component.tsx

```diff
@@ -0,0 +1,314 @@
+import {Component} from '@angular/core';
+import * as fp from 'lodash/fp';
+import * as React from 'react';
+import {validate} from 'validate.js';
+
+import {Button, Clickable} from 'app/components/buttons';
+import {ClrIcon} from 'app/components/icons';
+import {TextArea, TextInput, ValidationError} from 'app/components/inputs';
+import {TooltipTrigger} from 'app/components/popups';
+import {SpinnerOverlay} from 'app/components/spinners';
+import {profileApi} from 'app/services/swagger-fetch-clients';
+import colors from 'app/styles/colors';
+import {reactStyles, ReactWrapperBase, withUserProfile} from 'app/utils';
+import {Profile} from 'generated/fetch';
+
+
+const styles = reactStyles({
+  h1: {
+    color: colors.purple[0],
+    fontSize: 20,
+    fontWeight: 500,
+    lineHeight: '24px'
+  },
+  inputLabel: {
+    color: colors.purple[0],
+    fontSize: 14, lineHeight: '18px',
+    marginBottom: 6
+  },
+  inputStyle: {
+    width: 250,
+    marginRight: 20
+  },
+  longInputStyle: {
+    height: 175, width: 420,
+    resize: 'both'
+  },
+  box: {
+    backgroundColor: '#fff',
+    borderRadius: 8,
+    padding: 21
+  }
+});
+
+const labelSubstitutions = {
+  givenName: 'firstName',
+  familyName: 'lastName',
+  currentPosition: 'currentPosition',
+  organization: 'organization',
+  areaOfResearch: 'currentResearch'
+};
+const required = {presence: {allowEmpty: false}};
+const notTooLong = maxLength => ({
+  length: {
+    maximum: maxLength,
+    tooLong: 'must be %{count} characters or less'
+  }
+});
+const validators = {
+  firstName: {...required, ...notTooLong(80)},
+  lastName: {...required, ...notTooLong(80)},
+  currentPosition: {...required, ...notTooLong(255)},
+  organization: {...required, ...notTooLong(255)},
+  currentResearch: required,
+};
+
+export const ProfilePageReact = withUserProfile()
+(class ProfilePageReactComponent extends React.Component<
+  { profileState: {profile: Profile, reload: Function} },
+  { profileEdits: Profile, updating: boolean }
+> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      profileEdits: props.profileState.profile || {},
```

Comment:
Should this somehow be shallow-copied across? Or do we just hold the assumption that state will never be directly mutated?

Reply:
All updates are using `fp`, since they're deeper, so it's never mutated.